### PR TITLE
Improve count performance

### PR DESCRIPTION
**Pull request by bkamins**
This is a speculative PR as I do not fully understand why the original code was not properly optimized by the compiler (so maybe a better fix is to change compiler).

Here is an example of the performance difference:
```
julia> x = rand([1, missing], 10^6);

julia> @btime count(ismissing, $x)
  501.200 μs (0 allocations: 0 bytes)
500396

julia> @btime count(Base.Generator(ismissing, $x))
  70.800 μs (0 allocations: 0 bytes)
500396
```
(this is run on 1.6, but this PR is essentially doing the same by introducing a function barrier)

### CODE REVIEW

File: base/reduce.jl [hunk 1]
```diff
@@ -1200,10 +1200,12 @@ count(itr; init=0) = count(identity, itr; init)
 
 count(f, itr; init=0) = _simple_count(f, itr, init)
 
-function _simple_count(pred, itr, init::T) where {T}
+_simple_count(pred, itr, init) = _simple_count_helper(Generator(pred, itr), init)
+
+function _simple_count_helper(g, init::T) where {T}
     n::T = init
     for x in itr
```
Comment (bkamins) [inline base/reduce.jl:1207]:
```suggestion
    for x in g
```

**Comment by KristofferC** [comment]
Some notes, looking at:

```
@code_llvm Base._simple_count(ismissing, x, 0)
``` 
```
 @code_llvm Base._simple_count(identity, Base.Generator(ismissing, x), 0)
```

the second case manages to vectorized. The LLVM vectorization pass reports:

```
remark: reduce.jl:969:0: loop not vectorized: loop control flow is not understood by vectorizer
remark: reduce.jl:969:0: loop not vectorized: loop control flow is not understood by analyzer
remark: reduce.jl:969:0: loop not vectorized
```

in the first case but

```
remark: array.jl:777:0: vectorized loop (vectorization width: 4, interleaved count: 4)
```

in the second case.

File: base/reduce.jl [hunk 2]
```diff
@@ -1200,10 +1200,12 @@ count(itr; init=0) = count(identity, itr; init)
...
 
-function _simple_count(pred, itr, init::T) where {T}
+_simple_count(pred, itr, init) = _simple_count_helper(Generator(pred, itr), init)
+
+function _simple_count_helper(g, init::T) where {T}
     n::T = init
-    for x in itr
-        n += pred(x)::Bool
+    for x in g
+        n += x::Bool
```
Comment (KristofferC) [inline base/reduce.jl:1208]: Is this just `n += true` now?

Comment (bkamins) [reply base/reduce.jl:1208]: no `pred` can return `true` or `false`, e.g.:
```
julia> collect(Base.Generator(ismissing, [1,missing,2]))
3-element Vector{Bool}:
 0
 1
 0
```
and we still need `::Bool` assertion

Comment (KristofferC) [reply base/reduce.jl:1208]: Oh yeah, somehow I thought the first argument was a filter with this syntax.

Comment (bkamins) [reply base/reduce.jl:1208]: That is why I was surprised that the compiler does not vectorize it (the only thing I do is introduce `Generator` wrapper and add a function barrier - but both are needed to force vectorization)

**Comment by vtjnash** [comment]
This also fixes it. @Keno I think is about time we switched this to assuming subtyping and intersection are perfect, WDYT?
```diff
diff --git a/base/compiler/ssair/inlining.jl b/base/compiler/ssair/inlining.jl
index 0420d137da..328f1f0f3a 100644
--- a/base/compiler/ssair/inlining.jl
+++ b/base/compiler/ssair/inlining.jl
@@ -426,7 +426,7 @@ function ir_inline_item!(compact::IncrementalCompact, idx::Int, argexprs::Vector
     return_value
 end
 
-const fatal_type_bound_error = ErrorException("fatal error in type inference (type bound)")
+#const fatal_type_bound_error = ErrorException("fatal error in type inference (type bound)")
 
 function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int,
                                argexprs::Vector{Any}, linetable::Vector{LineInfoNode},
@@ -443,18 +443,20 @@ function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int,
         @assert !isa(metharg, UnionAll)
         cond = true
         @assert length(atype.parameters) == length(metharg.parameters)
-        for i in 1:length(atype.parameters)
-            a, m = atype.parameters[i], metharg.parameters[i]
-            # If this is always true, we don't need to check for it
-            a <: m && continue
-            # Generate isa check
-            isa_expr = Expr(:call, isa, argexprs[i], m)
-            ssa = insert_node_here!(compact, NewInstruction(isa_expr, Bool, line))
-            if cond === true
-                cond = ssa
-            else
-                and_expr = Expr(:call, and_int, cond, ssa)
-                cond = insert_node_here!(compact, NewInstruction(and_expr, Bool, line))
+        if !item.fully_covered || next_cond_bb != generic_bb
+            for i in 1:length(atype.parameters)
+                a, m = atype.parameters[i], metharg.parameters[i]
+                # If this is always true, we don't need to check for it
+                a <: m && continue
+                # Generate isa check
+                isa_expr = Expr(:call, isa, argexprs[i], m)
+                ssa = insert_node_here!(compact, NewInstruction(isa_expr, Bool, line))
+                if cond === true
+                    cond = ssa
+                else
+                    and_expr = Expr(:call, and_int, cond, ssa)
+                    cond = insert_node_here!(compact, NewInstruction(and_expr, Bool, line))
+                end
             end
         end
         insert_node_here!(compact, NewInstruction(GotoIfNot(cond, next_cond_bb), Union{}, line))
@@ -495,8 +497,8 @@ function ir_inline_unionsplit!(compact::IncrementalCompact, idx::Int,
     bb += 1
     # We're now in the fall through block, decide what to do
     if item.fully_covered
-        e = Expr(:call, GlobalRef(Core, :throw), fatal_type_bound_error)
-        insert_node_here!(compact, NewInstruction(e, Union{}, line))
+        #e = Expr(:call, GlobalRef(Core, :throw), fatal_type_bound_error)
+        #insert_node_here!(compact, NewInstruction(e, Union{}, line))
         insert_node_here!(compact, NewInstruction(ReturnNode(), Union{}, line))
         finish_current_bb!(compact, 0)
     else
```

**Comment by ViralBShah** [comment]
Bump.

**Comment by ViralBShah** [comment]
Would be nice to get this in.

**Comment by oscardssmith** [comment]
Is this just ready to merge?

**Comment by bkamins** [comment]
I assumed this should not be merged, but rather @vtjnash and @Keno would make a more general fix per the comment above and then this PR would be just closed as not needed.

**Comment by ViralBShah** [comment]
@bkamins Does this help with any of the recent DataFrames performance work and benchmarks? Trying to figure out if it is a nice to eventually have, or is it affecting performance sensitive things and should be addressed sooner.

**Comment by bkamins** [comment]
This does not affect DataFrames.jl directly. However, it affects any data science workflow (thus DataFrames.jl indirectly) as `missing` values are super common in these contexts. Another basic thing that has a similar problem is https://github.com/JuliaLang/julia/issues/40768 (but it is also non-critical).

What is holding writing https://github.com/JuliaLang/www.julialang.org/issues/1282 are indeed performance issues, but they are related to https://github.com/JuliaLang/julia/issues/40840. However, I fear that #40840 will not be resolved quickly (if at all; and even if this is resolved the fact that GC in Julia is not multi-threading friendly will keep hitting us constantly in data science workflows).

Therefore @quinnj is now rewriting CSV.jl to avoid GC issues as much as possible. After new CSV.jl will be out I expect that in https://h2oai.github.io/db-benchmark/ the only thing that we are currently bad (i.e. joins on large data sets) will be much better. Essentially in 5GB join tests we hopefully should go down from 402 seconds to ~100 seconds (without changing anything in DataFrames.jl).

Then we will be ready to make announcements (of course there will still be room for improvements but they will be marginal - in general data.table and Polars are already very well tuned in terms of performance, so we should not expect to be able to be much better, but we essentially already have a comparable performance; here https://opendatascience.com/saying-hello-to-dataframes-jl/ is a small comparison that is not GC bound, as I am using integers only there).

**Comment by ViralBShah** [comment]
Thanks @bkamins - Very useful.

**Comment by skanskan** [comment]
@bkamins I don't understand what CSV.jl has to do with the performance of dataframes.jl in  the  joining benchmarks.
Isn't CSV.jl just used to read and write csv files?

**Comment by bkamins** [comment]
I will talk about it during my presentation during JuliaCon2021 😄 (which is now accepted).

The chain of consequences is:
* CSV.jl currently reads in strings as `String` type
* `String` type is not interned in Julia; all other packages fast like data.table or Polars do not have this problem (they use different approaches to read-in the data)
* because it is not interned it takes part in garbage collection
* because of this every time GC is invoked it has to scan through all `String` objects in memory doing marking (this is super slow as there can be billions of such small objects)
* DataFrames.jl is currently multi-threaded, but garbage collection in Julia is single threaded, which means that when GC is invoked everything is frozen until garbage collection is finished

The consequence is that in H2O benchmarks of joins garbage collection is around 75% of all run-time of joining (before some changes that I did recently it was around 90% of all run time).

Now, how to resolve it:
* improve GC in Julia Base (this is something that Julia core devs are aware of and I hope at some point it will be done - but I do not expect this to happen fast);
* do not use `String` type to store strings;

And CSV.jl changes are exactly addressing the second point - @quinnj is developing changes that will provide two additional options of storing strings except for `String` type:
* storing short strings in-line (so this is kind of like string interning in R, but better as it fully avoids allocations, the limitation is that only short strings can be handled this way - but short strings are most common strings in practice anyway - so this is super useful to have)
* having one big allocated object and having individual strings be just pointers to its fragments (something like `SubString` type; this is an approach that e.g. Polars does as this roughly is the way Apache Arrow approaches to storing strings)

In short the answer is: `String` type is slowing us down because of various interlinked design decisions in Julia Base, so CSV.jl will avoid crating `String` objects to work around this issue.

Is this clearer now?

**Comment by PallHaraldsson** [comment]
>This also fixes it.

@vtjnash, is that other fix (that I did not review) merged? Either way, can the PR here be merged, as it seems simple and to give a good speedup, while less general. Given the "forgetmenot" label, I would like to see either or both in Julia 1.7.

**Comment by oscardssmith** [comment]
We are way passed the point of getting this in 1.7. That said, we absolutely should try to get this merged for 1.8 if it's a good improvement.

**Comment by bkamins** [comment]
If @Keno and @vtjnash can provide a general solution (I understood from the discussion above that this is the case) then I think this PR is not needed. Otherwise I think `count` is a very fundamental function so it is better to have a custom half-measure solution for it than not have it.

**Comment by timholy** [comment]
One general solution is to force specialization:
```julia
function _simple_count(pred::F, itr, init::T) where {F,T}
```
You will likely need to do this on `count` too though worth checking (these things change over time).

To be clear, changes like these give you better runtime performance at the expense of worse latency. Often that's worth it, but just FYI.

**Comment by oscardssmith** [comment]
Adding this to the triage label so we can get discussion as to whether we like this. IMO we should merge it.

**Comment by JeffBezanson** [comment]
Let's merge this for now, and open a new issue for the compiler fix.
